Update field mappings for ELSER pipelines

seanstory opened the pull request:
## Summary

When creating an ML inference pipeline that uses the new ELSER model, we need to be able to specify which fields to apply it to, and where the resulting values should go. These fields can then determine what mapping changes need to be made to the underlying index (setting `rank_features` fields).

This PR adds `field_mappings` as a param to our kibana endpoint for creating these ML Inference pipelines, and based on that parameter, attempts to update mappings.

A sample request might be:
```
curl -XPOST -u elastic:changeme "http://localhost:5601/mob/internal/enterprise_search/indices/search-test/ml_inference/pipeline_processors" \
-H 'kbn-xsrf: kibana' \
-H 'Content-Type: application/json' \
-d '{
  "field_mappings": {
    "input": "output"
  },
  "pipeline_definition": {
    "description" : "text",
    "processors": []
  },
  "pipeline_name": "test"
}' | jq
```

And sample output might be:
```
{
  "created": "ml-inference-test",
  "mapping_updated": true
}
```
OR
```
{
  "statusCode": 409,
  "error": "Conflict",
  "message": "One or more target fields for this pipeline already exist with a type that is incompatible with the specified model. Ensure that each target field is unique and not already in use.",
  "attributes": {
    "error_code": "mapping_update_failed"
  }
}
```

### Checklist


- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

seanstory commented:
WIP (still need to add tests), but from manual testing, it seems to be working. Figured I'd throw this up in case @demjened needs it sooner than later.

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,92 @@
+  }
+};
+
+const formELSERMappingProperties = (sourceFields: string[], targetFields: string[]) => {
```
Nit: I think acronyms in function names follow camelCase notation (see [Microsoft's style guide](https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)?redirectedfrom=MSDN) -> `HtmlButton`), but I'm not sure what our team's guidelines are for this.
```suggestion
const formElserMappingProperties = (sourceFields: string[], targetFields: string[]) => {
```

demjened reviewed (commented):
Good start! Please check the comments about refactoring and style suggestions.

seanstory commented on x-pack/plugins/enterprise_search/server/routes/enterprise_search/indices.ts:
```diff
@@ -463,9 +468,14 @@ export function registerIndexRoutes({
           client.asCurrentUser
         );
+        const mappingResponse = fieldMappings
+          ? (await updateMlInferenceMappings(indexName, fieldMappings, client.asCurrentUser))
```
👍 I didn't want it there because `createAndReferenceMlInferencePipeline` doesn't imply mapping change side effects, but I didn't think about just renaming the function. Yep, that works for me.

Though, will you need any indication that the mapping change was successful separate from the pipeline creation being successful? Right now having them as two distinct function calls makes it easy to have the response indicate two distinct successes. Still doable as 1 function, just not sure if you want that or not.

seanstory commented:
Still not finished (needs more tests than just the happy path) but @demjened I've addressed your first few comments, added _a_ test, and fixed existing failing tests. Should wrap up tomorrow 🤞

seanstory commented:
This PR should be ready for review now.

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,102 @@
+  fieldMappings: FieldMapping[],
+  esClient: ElasticsearchClient
+) => {
+  const sourceFields = fieldMappings.map((fieldMapping) => {
+    return fieldMapping.sourceField;
+  });
```
Minor: since this is a simple return statement, you can use destructuring:
```suggestion
  const sourceFields = fieldMappings.map(({ sourceField }) => sourceField);
```

You can read this as "while iterating on `fieldMappings` destructure each object: extract the `sourceField` property into a local variable of the same name (left side of `=>`). Then return that variable as-is (right side of `=>`)."

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,102 @@
+  });
+
+  const nonDefaultTargetFields = fieldMappings
+    .filter((fieldMapping) => {
+      return fieldMapping.targetField !== `ml.inference.${fieldMapping.sourceField}_expanded`;
+    })
+    .map((fieldMapping) => {
+      return fieldMapping.targetField;
+    });
```
I think in general we prefer to simplify single line return statements:
```suggestion
    .filter((fieldMapping) => fieldMapping.targetField !== `ml.inference.${fieldMapping.sourceField}_expanded)
    .map(({ sourceField }) => sourceField);
```

seanstory replied:
The first line was too long, so prettier still split it up. But I removed the `=> { return statement}` syntax in favor of `=> statement` as suggested.

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,102 @@
+  }
+};
+
+const formElserMappingProperties = (sourceFields: string[], targetFields: string[]) => {
```
Minor: Ah, after a closer look, now I understand the use of the word "form". 😅  I'd use "generate" instead - I associate "form" with the UI component that takes the input; it might be confusing to others too.

demjened replied:
Another minor: in the frontend code we tend to refer to ELSER inference as ["text expansion"](https://github.com/search?q=repo%3Aelastic%2Fkibana%20text_expansion&type=code), since that's the name of the ML model task and the condition for such special logic. ELSER is more of a marketing name and it's used on UI labels only.

seanstory replied:
switched to "generate" and "text expansion"

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,102 @@
+
+  const nonDefaultTargetFields = fieldMappings
+    .filter((fieldMapping) => {
+      return fieldMapping.targetField !== `ml.inference.${fieldMapping.sourceField}_expanded`;
```
I'm not sure about this filter. We do want to put `targetField` mappings for every selected `sourceField`, and for now it's always `ml.inference.${fieldMapping.sourceField}_expanded`.
Wouldn't this filter exclude all target fields from the mapping update statement?

seanstory replied:
Yes. If you keep reading, the `nonDefaultTargetFields` end up being `[]` for now, but the right mappings are getting generated from the `sourceFields`.

This is because I'd originally designed this expecting that the user could specify custom targets in 8.8. Even though they can't, eventually they will be able to, and this code will be ready for that.

demjened commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/pipeline_processors/create_ml_inference_pipeline.ts:
```diff
@@ -56,9 +61,15 @@ export const createAndReferenceMlInferencePipeline = async (
   return {
-    ...createPipelineResult,
-    addedToParentPipeline: addSubPipelineResult.addedToParentPipeline,
+    added_to_parent_pipeline: addSubPipelineResult.addedToParentPipeline,
+    created_pipeline: createPipelineResult.created,
+    mapping_updated: mappingResponse,
+    pipeline_id: createPipelineResult.id,
```
Looks like we're changing the return type. Isn't this breaking [routes/enterprise_search/indices](https://github.com/elastic/kibana/pull/154291/files#diff-f9440a95c3204a819bc8f1fe51dafe281544bedc7083e0315795c5dc598cd933R463-R472)?

seanstory replied:
Nope. The return of the API was returning the value `id` before and now is returning the value `pipeline_id`, which are the same value, just renamed property. And the returned response's key for this ID is not changing: https://github.com/elastic/kibana/pull/154291/files#diff-f9440a95c3204a819bc8f1fe51dafe281544bedc7083e0315795c5dc598cd933R475-R477

demjened reviewed (commented):
@seanstory  Please take a look at the minor suggestions and questions, especially the one about the modified return type.

kibanamachine commented:
merge conflict between base and head

demjened reviewed (approved):
LGTM

TattdCodeMonkey commented on x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/update_ml_inference_mappings.ts:
```diff
@@ -0,0 +1,99 @@
+  fieldMappings: FieldMapping[],
+  esClient: ElasticsearchClient
+) => {
+  const sourceFields = fieldMappings.map(({ sourceField }) => sourceField);
+
+  const nonDefaultTargetFields = fieldMappings
+    .filter(
+      (fieldMapping) =>
+        fieldMapping.targetField !== `ml.inference.${fieldMapping.sourceField}_expanded`
+    )
+    .map((fieldMapping) => fieldMapping.targetField);
+
+  // Today, we only update mappings for text_expansion fields.
+  const mapping = generateTextExpansionMappingProperties(sourceFields, nonDefaultTargetFields);
```
Should we combine this logic parsing `fieldMappings` to `mapping` and unit test it? 

Not sure if we are concerned about any possible edge cases where a `fieldMappings` value could result in `mapping` with no source or target fields. I don't think it's currently possible given the usage of the endpoint but looking at the code doing a filter it strikes me as technical possibility. Not sure if its worth being concerned about or not though.

seanstory replied:
And this is why I shouldn't enable auto-merge. :( Sorry Rodney, to have effectively ignored this comment.
If there are no fieldMappings, then no `properties` with a `type` would be defined, so the putMapping request would effectively no-op. I'm not too worried about that edge case.

There are unit tests to validate that these mappings are created correctly, though: https://github.com/elastic/kibana/pull/154291/files/2833e24536e78a7701d531435ea12be1a475984e#diff-b698a365727af822707e5c4ddea43e134500def62ea5e86167b0c59c27c9430fR23-R52

TattdCodeMonkey reviewed (approved):
One comment that I don't think is a blocker
